### ISSUE TITLE
Is there a way to implement a RWLock ontop of async-mutex?

### ISSUE THREAD

**Issue by CMCDragonkai** [DirtyHairy/async-mutex#48]
Seems like https://en.wikipedia.org/wiki/Readers%E2%80%93writer_lock could be done on top of async-mutex.

**Comment by DirtyHairy** [comment]
A RW lock is a possibility, but would most likely by a separate mutex class. Do you have a usecase in mind?

**Comment by CMCDragonkai** [comment]
This was my attempt using `async-mutex`. It's read-preferring.

**Comment by CMCDragonkai** [comment]
https://gist.github.com/CMCDragonkai/4de5c1526fc58dac259e321db8cf5331

**Comment by CMCDragonkai** [comment]
I was wondering how to do a write-preferring RWLock, the wikipedia article gives an method using condition variables, but we don't have this here.

A write-preferring one should mean that when a write call is done, it should block and queue up all the reads.

**Comment by DirtyHairy** [comment]
That's an interesting exercise 😏 I *think* the following algorithm with two mutexes and two counters should work:

Reader:

1. Lock write mutex
2. Check read counter, lock read mutex if it is zero
3. Increment read counter
4. Release write mutex
5. ... [do async stuff]
6. Decrement read counter
7. Check read counter,  release read mutex if it is zero

Writer:

1. Check write counter, lock write mutex if it is zero
2. Increment write counter
3. Lock read mutex
4. ... [do async stuff]
5. Release read mutex
6. Decrement write counter
7. Check write counter, release write mutex if it zero

**Comment by CMCDragonkai** [comment]
I've written a write-preferring RWLock. https://gist.github.com/CMCDragonkai/4de5c1526fc58dac259e321db8cf5331

**Comment by eblocha** [comment]
Awesome work, Dragonkai. One small problem I had was that readers could skip ahead of writers that called acquireRead first. I made some modifications using the [double-ended-queue](https://github.com/petkaantonov/deque) package.

Essentially, I added a deque to keep track of the order of readers and writers, and forced readers to wait if a writer has requested the lock. Then, readers will unlock if the next queued item is a writer, or the queue is empty.

This gist has the modifications (I only did the read-preferring lock, as well as a test case to illustrate the issue).

https://gist.github.com/eblocha/6a84b4c8060a90fde8112e4cb5d2cdcd
